---
title: "README"
author: "Rasmus Kirkegaard"
date: "`r format(Sys.time(), '%d %B, %Y')`"
output: github_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
options(scipen = 10)
```

# R10.4.1 Zymo HMW basecalling


## Aim

To check the quality of nanopore data with the ZymoHMW mock DNA.

## Conclusion
There are some serious improvements in raw read accuracy between the fast, hac and sup models.

## Data availability

[https://github.com/Kirk3gaard/MicroBench](https://github.com/Kirk3gaard/MicroBench)


```{r load_libraries_and_data,echo=FALSE,warning=FALSE,message=FALSE}
library(tidyverse)
library(gridExtra)
tt2 <- ttheme_default(core=list(fg_params=list(hjust=1, x = 0.95, fontsize = 10)),
                      colhead=list(fg_params=list(fontsize = 12)))

# duplex read QC new ref
get_read_QC<-function(model="dna_r10.4.1_e8.2_400bps_hac@v4.0.0") {
  files<-dir(pattern = paste0(model,"_ref_E.coli_.*.NanoPlot-data.tsv.gz"),path = "temp/",full.names = T)
  NP_QC_ref_Escherichia_coli <- files %>% map(.,~read_tsv(., show_col_types = FALSE)) %>% reduce(rbind) %>% mutate(REF="Escherichia_coli")
  files<-dir(pattern = paste0(model,"_ref_L.monocytogenes_.*.NanoPlot-data.tsv.gz"),path = "temp/",full.names = T)
  NP_QC_ref_Listeria_monocytogenes<-files %>% map(.,~read_tsv(., show_col_types = FALSE)) %>% reduce(rbind) %>% mutate(REF="Listeria_monocytogenes")
  files<-dir(pattern = paste0(model,"_ref_P.aeruginosa_.*.NanoPlot-data.tsv.gz"),path = "temp/",full.names = T)
  NP_QC_ref_Pseudomonas_aeruginosa<- files %>% map(.,~read_tsv(., show_col_types = FALSE)) %>% reduce(rbind) %>% mutate(REF="Pseudomonas_aeruginosa")
  files<-dir(pattern = paste0(model,"_ref_B.subtilis_.*.NanoPlot-data.tsv.gz"),path = "temp/",full.names = T)
  NP_QC_ref_Bacillus_subtilis<- files %>% map(.,~read_tsv(., show_col_types = FALSE)) %>% reduce(rbind) %>% mutate(REF="Bacillus_subtilis")
  files<-dir(pattern = paste0(model,"_ref_S.enterica_.*.NanoPlot-data.tsv.gz"),path = "temp/",full.names = T)
  NP_QC_ref_Salmonella_enterica<-files %>% map(.,~read_tsv(., show_col_types = FALSE)) %>% reduce(rbind) %>% mutate(REF="Salmonella_enterica")
  files<-dir(pattern = paste0(model,"_ref_E.faecalis_.*.NanoPlot-data.tsv.gz"),path = "temp/",full.names = T)
  NP_QC_ref_Enterococcus_faecalis<-files %>% map(.,~read_tsv(., show_col_types = FALSE)) %>% reduce(rbind) %>% mutate(REF="Enterococcus_faecalis")
  files<-dir(pattern = paste0(model,"_ref_S.aureus_.*.NanoPlot-data.tsv.gz"),path = "temp/",full.names = T)
  NP_QC_ref_Staphylococcus_aureus<-files %>% map(.,~read_tsv(., show_col_types = FALSE)) %>% reduce(rbind) %>% mutate(REF="Staphylococcus_aureus")
  NP_QC_combined<-bind_rows(NP_QC_ref_Bacillus_subtilis,
                            NP_QC_ref_Enterococcus_faecalis,
                            NP_QC_ref_Escherichia_coli,
                            NP_QC_ref_Listeria_monocytogenes,
                            NP_QC_ref_Pseudomonas_aeruginosa,
                            NP_QC_ref_Salmonella_enterica,
                            NP_QC_ref_Staphylococcus_aureus) %>% 
    mutate(model=model)
return(NP_QC_combined)
}

NP_QC_fast_5.0.0<-get_read_QC(model = "PAS01578.dorado0.7.3.bmdna_r10.4.1_e8.2_400bps_5.0.0_fast.sim")
NP_QC_hac_5.0.0<-get_read_QC(model = "PAS01578.dorado0.7.3.bmdna_r10.4.1_e8.2_400bps_5.0.0_hac.sim")
NP_QC_hacdup_5.0.0<-get_read_QC(model = "PAS01578.dorado0.7.3.bmdna_r10.4.1_e8.2_400bps_5.0.0_hac.dup")
NP_QC_sup_5.0.0<-get_read_QC(model = "PAS01578.dorado0.7.3.bmdna_r10.4.1_e8.2_400bps_5.0.0_sup.sim")
NP_QC_supdup_5.0.0<-get_read_QC(model = "PAS01578.dorado0.7.3.bmdna_r10.4.1_e8.2_400bps_5.0.0_sup.dup")
NP_QC_combined<-bind_rows(NP_QC_fast_5.0.0,
                          NP_QC_hac_5.0.0,
                          NP_QC_hacdup_5.0.0,
                          NP_QC_sup_5.0.0,
                          NP_QC_supdup_5.0.0) %>%
  mutate(model_version=gsub(x = model,pattern=".*bps_([0-9].[0-9].[0-9])_.*",replacement = "\\1"),
         basecalling_model=gsub(x = model,pattern=".*_(.*)",replacement = "\\1"))

### read coverage
draft_assemblies<-list.files(path = "results/",pattern = "*.assembly_info.txt",full.names = T)
cov_tab<-tibble()
for (f in draft_assemblies) {
  cov_temp<-read_tsv(file = f,show_col_types = FALSE) %>% mutate(draft=f)
  cov_tab<-bind_rows(cov_tab,cov_temp)
}
cov_tab<-cov_tab %>% mutate(contigName=`#seq_name`,coverage=cov.,circular=circ.,
                            ReadSet=str_replace(string =draft,pattern = "results//(.*dna.*).assembly_info.txt",replacement = "\\1")) %>% 
  select(ReadSet,contigName,coverage,circular)

### Assembly based stats
fastani<-read_tsv(file = "results/fastani.tsv",col_names = c("query","ref","ANI","aligned_segments","total_segments",show_col_types = FALSE)) %>% 
  mutate(Assembly=str_replace(string = query,pattern = "results/bins/(.*).fa.*",replacement = "\\1"),
         ref=str_replace(string = ref,pattern = ".*/(.*).fasta",replacement = "\\1")) %>% 
  select(Assembly,ref,ANI)

ref_genomes=list.files(path = "results/",pattern = "quast_.*.tsv",full.names = T)

# Load quast data
quast_tab<-tibble()
for (f in ref_genomes) {
  quast_temp<-read_tsv(file = f,show_col_types = FALSE) %>% mutate(ref=f) %>% filter(`Total aligned length`!="-")
  quast_tab<-bind_rows(quast_tab,quast_temp)
}
quast_tab<-quast_tab %>% mutate(ref=str_replace(string = ref,pattern = ".*quast_(.*).tsv",replacement = "\\1"),
                                Assembly=str_trim(Assembly)) %>% filter(as.numeric(`Total aligned length`)>10^6) %>%
  select(Assembly,ref,`Total length`,`Reference length`,`Largest alignment`,`Total aligned length`,`# indels per 100 kbp`,`# mismatches per 100 kbp`) %>%
  mutate(`Total aligned length`=as.numeric(`Total aligned length`),
         `# indels per 100 kbp`=as.numeric(`# indels per 100 kbp`),
         `# mismatches per 100 kbp`=as.numeric(`# mismatches per 100 kbp`)) %>%
  filter(Assembly!=ref)

###
genome_stats<-fastani %>% right_join(quast_tab,by = c("Assembly", "ref")) %>%
  mutate(contigName=str_replace(string = Assembly,pattern = ".*(contig.*)",replacement = "\\1"),
         ReadSet=str_replace(string = Assembly,pattern = "(.*).flye.*",replacement = "\\1")) %>%
  filter(ANI>95) %>%
  left_join(cov_tab,by = c("contigName", "ReadSet")) %>%
  mutate(asmtype=str_replace(string = Assembly,pattern = ".*(fly.*).contig.*",replacement = "\\1")) %>%
  mutate(model=str_replace(ReadSet,pattern = "-.*",replacement = "")) %>%
  mutate(ref=str_replace(ref,pattern = "_hifiasm",replacement = "")) %>%
  filter(ref!="S.cerevisiae") %>%
  mutate(model_version=gsub(x = model,pattern=".*bps_([0-9].[0-9].[0-9])_.*",replacement = "\\1"),
         basecalling_model=gsub(x = model,pattern=".*_(.*)",replacement = "\\1"))
```

## NP reads mapped to the refs overall (percent identity)

```{r fig.width=10,fig.height=7,echo=FALSE,warning=FALSE,message=FALSE}
NP_QC_combined %>% ggplot(aes(x = percentIdentity,col=basecalling_model))+geom_density()+geom_vline(xintercept = 99,col="red",linetype="dashed")+scale_x_continuous(breaks = 85:100)+coord_cartesian(xlim=c(85, 100))+xlab("Percent identity")+facet_wrap(~model_version,ncol = 1,scales = "free_y")
```

## Longest perfect read

```{r fig.width=10,fig.height=30,echo=FALSE,warning=FALSE,message=FALSE}
order_df<-NP_QC_combined %>% group_by(model) %>% filter(percentIdentity==100) %>% filter(aligned_lengths==max(aligned_lengths)) %>% select(model,aligned_lengths)
order_vector<-order_df$model[order(order_df$aligned_lengths)]

NP_QC_combined %>% group_by(basecalling_model,model_version,REF) %>% filter(percentIdentity==100) %>% filter(aligned_lengths==max(aligned_lengths)) %>% 
  mutate(model=factor(model,levels=order_vector)) %>%
  ggplot(aes(x=lengths,y=model,label=lengths,col=basecalling_model))+geom_point()+facet_wrap(~REF,ncol = 1)+geom_text(nudge_x = 5000)
```

## NP reads mapped to the refs overall (phred scale)

Phred scores for perfect matching reads are calculated as recommended by [Armin Topfer](https://twitter.com/kirk3gaard/status/1397457000217423873) which takes length into account by adjusting the percent identity value 100*(1-1/(length+1)). This implies that a perfect read of 100 bp will achieve a phred score of ~20, while 1000 bp caps out at ~30 etc. So forget about your short reads being Q40. 

```{r fig.width=10,fig.height=7,echo=FALSE,warning=FALSE,message=FALSE}
PHREDNORMPERFECT=function(length=10) {
  adj_percentidentity<-100*(1-1/(length+1))
  phred<--10*log10((100-adj_percentidentity)/100)
  return(phred)
}
NP_QC_combined %>% 
  mutate(Phred=case_when(percentIdentity==100~PHREDNORMPERFECT(length=lengths),
                         TRUE~-10*log10((100-percentIdentity)/100))) %>%
  ggplot(aes(x = Phred,col=basecalling_model))+geom_density()+geom_vline(xintercept = 20,col="red",linetype="dashed")+xlab("Phred score (Q)")+facet_wrap(~model_version,ncol = 1)
```

As the phred score calculation above somewhat hides the fraction of perfect reads here they come as such. 

```{r fig.width=10,fig.height=7,echo=FALSE,warning=FALSE,message=FALSE}
NP_QC_combined %>% 
  mutate(Phred=case_when(percentIdentity==100~50,
                         TRUE~-10*log10((100-percentIdentity)/100))) %>%
  ggplot(aes(x = Phred,col=basecalling_model))+geom_density()+geom_vline(xintercept = 20,col="red",linetype="dashed")+xlab("Phred score (Q)")+facet_wrap(~model_version,ncol = 1)+scale_x_continuous(labels=c("0" = "0", "10" = "10", "20" = "20", "30" = "30", "40" = "40", "50" = "Perfect"))+annotate("rect", xmin = 45, xmax = 47, ymin = 0, ymax = 0.4,
           alpha = 1,fill = "lightgray")
```

## all modes

### Indel rate vs coverage

```{r fig.width=10, echo=FALSE,warning=FALSE,message=FALSE}
genome_stats %>% 
  ggplot(aes(x=coverage,y = `# indels per 100 kbp`,col=model,shape=asmtype))+geom_point()+geom_line()+facet_wrap(~ref)+scale_y_log10()+coord_cartesian(xlim=c(0, 100))
```

### Mismatch rate vs coverage

```{r fig.width=10, echo=FALSE,warning=FALSE,message=FALSE}
genome_stats %>% 
  ggplot(aes(x=coverage,y = `# mismatches per 100 kbp`,col=model,shape=asmtype))+geom_point()+geom_line()+facet_wrap(~ref)+scale_y_log10()+coord_cartesian(xlim=c(0, 100))
```

## Fast mode

### Indel rate vs coverage

```{r fig.width=10, echo=FALSE,warning=FALSE,message=FALSE}
genome_stats %>% 
    filter(basecalling_model=="fast.sim") %>%
  ggplot(aes(x=coverage,y = `# indels per 100 kbp`,col=model,shape=asmtype))+geom_point()+geom_line()+facet_wrap(~ref)+scale_y_log10()+coord_cartesian(xlim=c(0, 100))
```

### Mismatch rate vs coverage

```{r fig.width=10, echo=FALSE,warning=FALSE,message=FALSE}
genome_stats %>% 
  filter(basecalling_model=="fast.sim") %>%
  ggplot(aes(x=coverage,y = `# mismatches per 100 kbp`,col=model,shape=asmtype))+geom_point()+geom_line()+facet_wrap(~ref)+scale_y_log10()+coord_cartesian(xlim=c(0, 100))
```

## HAC mode

### Indel rate vs coverage

```{r fig.width=10, echo=FALSE,warning=FALSE,message=FALSE}
genome_stats %>% 
  filter(basecalling_model=="hac.sim") %>%
  ggplot(aes(x=coverage,y = `# indels per 100 kbp`,col=model,shape=asmtype))+geom_point()+geom_line()+facet_wrap(~ref)+scale_y_log10()+coord_cartesian(xlim=c(0, 100))
```

### Mismatch rate vs coverage

```{r fig.width=10, echo=FALSE,warning=FALSE,message=FALSE}
genome_stats %>% 
  filter(basecalling_model=="hac.sim") %>%
  ggplot(aes(x=coverage,y = `# mismatches per 100 kbp`,col=model,shape=asmtype))+geom_point()+geom_line()+facet_wrap(~ref)+scale_y_log10()+coord_cartesian(xlim=c(0, 100))
```

## SUP mode

### Indel rate vs coverage

```{r fig.width=10, echo=FALSE,warning=FALSE,message=FALSE}
genome_stats %>% 
  filter(basecalling_model=="sup.sim") %>%
  ggplot(aes(x=coverage,y = `# indels per 100 kbp`,col=model,shape=asmtype))+geom_point()+geom_line()+facet_wrap(~ref)+scale_y_log10()+coord_cartesian(xlim=c(0, 100))
```

### Mismatch rate vs coverage

```{r fig.width=10, echo=FALSE,warning=FALSE,message=FALSE}
genome_stats %>% 
  filter(basecalling_model=="sup.sim") %>%
  ggplot(aes(x=coverage,y = `# mismatches per 100 kbp`,col=model,shape=asmtype))+geom_point()+geom_line()+facet_wrap(~ref)+scale_y_log10()+coord_cartesian(xlim=c(0, 100))
```

## Duplex mode

### Indel rate vs coverage

```{r fig.width=10, echo=FALSE,warning=FALSE,message=FALSE}
genome_stats %>% 
  filter(basecalling_model %in% c("hac.dup","sup.dup")) %>%
  ggplot(aes(x=coverage,y = `# indels per 100 kbp`,col=model,shape=asmtype))+geom_point()+geom_line()+facet_wrap(~ref)+scale_y_log10()+coord_cartesian(xlim=c(0, 100))
```

### Mismatch rate vs coverage

```{r fig.width=10, echo=FALSE,warning=FALSE,message=FALSE}
genome_stats %>% 
  filter(basecalling_model %in% c("hac.dup","sup.dup")) %>%
  ggplot(aes(x=coverage,y = `# mismatches per 100 kbp`,col=model,shape=asmtype))+geom_point()+geom_line()+facet_wrap(~ref)+scale_y_log10()+coord_cartesian(xlim=c(0, 100))
```


## Materials and methods

Here is a brief description of the tools used. For the exact commands check out the **Snakefile** in this repository ([Snakemake](https://snakemake.readthedocs.io/en/stable/) v. 7.18.2).

### DNA sequencing
DNA sample was the [Zymo Mock HMW standard](https://zymoresearch.eu/products/zymobiomics-hmw-dna-standard). The DNA was prepared for sequencing using the nanopore ligation sequencing kit (SQK-LSK114) and sequenced on a R10.4.1 nanopore PromethION flowcell (FLO-PRO114HD) with the "400 bp/s" mode (5khz sampling).

### Basecalling
#### 5 khz High Duplex PromethION data
The reads were basecalled using [dorado](https://github.com/nanoporetech/dorado) (v. 0.7.3) with fast, hac and sup accuracy mode using the 5.0.0 models.

### Read QC
Reads were mapped to the updated zymo reference genomes (hopefully goes public soon) using [minimap2](https://github.com/lh3/minimap2) (v. 2.24), and QC information was obtained using [NanoPlot](https://github.com/wdecoster/NanoPlot) (v. 1.41.0).

### Assembly
The reads were subsampled using [seqtk](https://github.com/lh3/seqtk) (v. 1.3) and assembled using [flye](https://github.com/fenderglass/Flye) (v. 2.9.5).

### Genome quality assessment
The assembled contigs were compared to the reference contigs using [QUAST](https://github.com/ablab/quast) (v. 5.2.0) and [fastANI](https://github.com/ParBLiSS/FastANI) (v. 1.33).